### PR TITLE
Make actor's tokio runtime configurable.

### PR DESCRIPTION
**Pull request by fulmicoton**
This PR is part of the effort of #1536,
handling of multiple indexes.
    
It changes the actor API to allow them to target
a specific tokio runtime.
    
After this PR, blocking actors are executed
in a shared tokio runtime, drastically reducing the
number of threads, and relies on cooperative
task yielding instead of multithreading for
concurrency.

### CODE REVIEW

File: quickwit-actors/src/spawn_builder.rs [hunk 1]
```diff
@@ -83,15 +91,149 @@ impl<A: Actor> SpawnBuilder<A> {
...
+        let span = actor.span(&ctx);
+        let loop_async_actor_future =
+            async move { actor_loop(actor, inbox, ctx, state_tx).await }.instrument(span);
+        let join_handle = runtime_handle.spawn(loop_async_actor_future);
+        let actor_handle = ActorHandle::new(state_rx, join_handle, ctx_clone);
         (mailbox, actor_handle)
     }
 }
+
+fn process_command<A: Actor>(
```
Comment (fulmicoton) [inline quickwit-actors/src/spawn_builder.rs:109]: that's actually more or less a copy paste from the runner.rs file.

File: quickwit-indexing/src/actors/merge_planner.rs [hunk 2]
```diff
@@ -53,10 +51,6 @@ impl Actor for MergePlanner {
         QueueCapacity::Bounded(0)
     }
 
-    fn runner(&self) -> ActorRunner {
-        ActorRunner::DedicatedThread
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/merge_planner.rs:57]: no clue why we run that as a blocking actor.

File: quickwit-indexing/src/actors/packager.rs [hunk 3]
```diff
@@ -102,15 +102,15 @@ impl Actor for Packager {
     }
 
     fn queue_capacity(&self) -> QueueCapacity {
-        QueueCapacity::Bounded(0)
+        QueueCapacity::Bounded(1)
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/packager.rs:105]: This had an impact on indexing throughput.

File: quickwit-actors/src/actor_handle.rs [hunk 4]
```diff
@@ -188,7 +188,13 @@ impl<A: Actor> ActorHandle<A> {
 
     /// Waits until the actor exits by itself. This is the equivalent of `Thread::join`.
     pub async fn join(self) -> (ActorExitStatus, A::ObservableState) {
-        let exit_status = self.join_handle.join().await;
+        let exit_status = self.join_handle.await.unwrap_or_else(|join_err| {
+            if join_err.is_panic() {
+                ActorExitStatus::Panicked
+            } else {
+                ActorExitStatus::Killed
```
Comment (fmassot) [inline quickwit-actors/src/actor_handle.rs:195]: just for your info, in tokio, the join error is either panic or a cancellation. Killed seems a bit harsh :)

File: quickwit-cli/src/main.rs [hunk 5]
```diff
@@ -116,6 +117,34 @@ async fn jemalloc_metrics_loop() -> tikv_jemalloc_ctl::Result<()> {
     }
 }
 
+/// If a bunch of tokio runtimes need to be started for actors,
+/// return the right configuration.
+///
+/// TODO making it configurable could be useful in the
```
Comment (fmassot) [inline quickwit-cli/src/main.rs:123]: `/// TODO making it configurable could be useful in the` future?

Comment (fmassot) [reply quickwit-cli/src/main.rs:123]: ok repetition at the end, so you can ditch "in the".

File: quickwit-cli/src/main.rs [hunk 6]
```diff
@@ -116,6 +117,34 @@ async fn jemalloc_metrics_loop() -> tikv_jemalloc_ctl::Result<()> {
...
+        }
+        CliCommand::Index(_) => Some(RuntimesConfiguration::default()),
+        CliCommand::Split(_) | CliCommand::Source(_) => None,
+    }
+}
+
+fn start_actor_runtimes(cli_command: &CliCommand) -> anyhow::Result<()> {
+    if let Some(runtime_configuration) = runtime_configuration_for_cmd(cli_command) {
+        quickwit_common::runtimes::initialize_runtimes(runtime_configuration)
+            .context("Failed to start runtimes")?;
```
Comment (fmassot) [inline quickwit-cli/src/main.rs:143]: I like the dot at the end.

File: quickwit-common/src/runtimes.rs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    fn light_for_tests() -> RuntimesConfiguration {
+        RuntimesConfiguration {
+            num_threads_blocking: 1,
+            num_threads_non_blocking: 1,
+        }
+    }
+
+    pub fn with_num_cpus(num_cpus: usize) -> Self {
+        let num_threads_non_blocking = if num_cpus > 6 { 2 } else { 1 };
+        let num_threads_blocking = (num_cpus - num_threads_non_blocking).max(1);
```
Comment (fmassot) [inline quickwit-common/src/runtimes.rs:70]: The rationale here is that we want to give the maximum number of threads to the CPU-consuming actors: Indexer, Merger, Packager.
You can add a comment maybe to justify the default values?

File: quickwit-indexing/src/actors/packager.rs (hunk 3 above)
Comment (fmassot) [reply quickwit-indexing/src/actors/packager.rs:105]: On the contrary, we will consume more RAM.

Comment (fulmicoton) [reply quickwit-indexing/src/actors/packager.rs:105]: Yes I saw a real difference in throughput. I don't quite understand why to be honest. 
Probably some race conditions due to the limitation of the thread pool.

(Indexer just needs to pass this value to the packager to yield while Packager is waiting for someone to yield.)

File: quickwit-actors/src/actor_handle.rs (hunk 4 above)
Comment (fulmicoton) [reply quickwit-actors/src/actor_handle.rs:195]: Killed is correct.

Names are for the client of our API. 
We want a name that matches quickwit's naming here, not tokio's naming.

This task being cancelled can only happen if an actor got killed.

File: quickwit-cli/src/main.rs (hunk 5 above)
Comment (fulmicoton) [reply quickwit-cli/src/main.rs:123]: fixed

**Comment by fulmicoton** [comment]
> LGTM. I'm only worried about the memory consumption with the change in the packager queue capacity. But I don't have any figures to justify my feeling.

If we have a single pipeline the usage is higher indeed.
If we have several the usage is already much better than before.

**Pull request merged**